hack/generated/controllers/crd_networking_nsg_test.go

```diff
@@ -0,0 +1,140 @@
+		Spec: network.NetworkSecurityGroups_Spec{
+			Location: tc.AzureRegion,
+			Owner:    testcommon.AsOwner(rg.ObjectMeta),
+		},
+	}
+
+	tc.CreateResourceAndWait(nsg)
+
+	tc.Expect(nsg.Status.Id).ToNot(BeNil())
+	armId := *nsg.Status.Id
+
+	// Perform a simple patch
+	old := nsg.DeepCopy()
+	nsg.Spec.Tags = map[string]string{
+		"foo": "bar",
+	}
+	tc.Patch(old, nsg)
+
+	objectKey := client.ObjectKeyFromObject(nsg)
+
+	// ensure state got updated in Azure
+	tc.Eventually(func() map[string]string {
+		updated := &network.NetworkSecurityGroup{}
+		tc.GetResource(objectKey, updated)
+		return updated.Status.Tags
+	}).Should(HaveKey("foo"))
+
+	// Run sub tests
+	tc.RunParallelSubtests(
+		testcommon.Subtest{
+			Name: "SecurityRules CRUD",
+			Test: func(testContext testcommon.KubePerTestContext) {
+				NetworkSecurityGroup_SecurityRules_CRUD(testContext, nsg.ObjectMeta)
+			},
+		},
+	)
+
+	tc.DeleteResourceAndWait(nsg)
+
+	// Ensure that the resource was really deleted in Azure
+	exists, retryAfter, err := tc.AzureClient.HeadResource(tc.Ctx, armId, string(network.NetworkSecurityGroupsSecurityRulesSpecAPIVersion20201101))
+	tc.Expect(err).ToNot(HaveOccurred())
+	tc.Expect(retryAfter).To(BeZero())
+	tc.Expect(exists).To(BeFalse())
+}
+
+func NetworkSecurityGroup_SecurityRules_CRUD(tc testcommon.KubePerTestContext, nsg metav1.ObjectMeta) {
+	rule1 := &network.NetworkSecurityGroupsSecurityRule{
+		ObjectMeta: tc.MakeObjectMeta("rule1"),
+		Spec: network.NetworkSecurityGroupsSecurityRules_Spec{
+			Owner:                    testcommon.AsOwner(nsg),
+			Protocol:                 network.SecurityRulePropertiesFormatProtocolTcp,
+			SourcePortRange:          to.StringPtr("23-45"),
+			DestinationPortRange:     to.StringPtr("46-56"),
+			SourceAddressPrefix:      to.StringPtr("*"),
+			DestinationAddressPrefix: to.StringPtr("*"),
+			Access:                   network.SecurityRulePropertiesFormatAccessAllow,
+			Priority:                 123,
+			Direction:                network.SecurityRulePropertiesFormatDirectionInbound,
+			Description:              to.StringPtr("The first rule of networking is don't talk about networking"),
+		},
+	}
+
+	rule2 := &network.NetworkSecurityGroupsSecurityRule{
+		ObjectMeta: tc.MakeObjectMeta("rule2"),
+		Spec: network.NetworkSecurityGroupsSecurityRules_Spec{
+			Owner:    testcommon.AsOwner(nsg),
+			Protocol: network.SecurityRulePropertiesFormatProtocolTcp,
+			SourcePortRanges: []string{
+				"23-45",
+				"5000-5100",
+			},
+			DestinationPortRange:     to.StringPtr("*"),
+			SourceAddressPrefix:      to.StringPtr("*"),
+			DestinationAddressPrefix: to.StringPtr("*"),
+			Access:                   network.SecurityRulePropertiesFormatAccessDeny,
+			Priority:                 124,
+			Direction:                network.SecurityRulePropertiesFormatDirectionInbound,
+		},
+	}
+
+	tc.CreateResourcesAndWait(rule1, rule2)
+	defer tc.DeleteResourceAndWait(rule1)
```

Comment:
if we have a `CreateResources` we should have a `DeleteResources`

Reply:
Fixed